### PR TITLE
Added Entitlement field + code-documentation

### PR DESCRIPTION
**Pull request by The-LukeZ**
added the `consumed` field for Entitlements + added a Doc-String for every Entitlement-Attribute.

## Summary by Sourcery

This pull request introduces a new 'consumed' field to the Entitlement class and adds comprehensive doc-strings for each attribute to improve code documentation.

* **New Features**:
    - Added a new 'consumed' field to the Entitlement class to indicate whether consumable items have been consumed.
* **Documentation**:
    - Added doc-strings for each attribute in the Entitlement class to provide inline documentation.

### CODE REVIEW

**Comment by mccoderpy** [comment]
That's not how docstrings are made in this lib. Please place them at the top of the method/class, otherwise autodoc will fail to read them.
You can take a look at the existing code for examples. (I should create a commit guildline 😅)

Also please consider updating the typing for the "new" API fields in `types/monetization.py`

PS: if you have general suggestions regarding improvement of in-code documentation for IDEs feel free to share them in a new issue

**Comment by The-LukeZ** [comment]
oh okay, I can't think of other "new" API fields besides the types of Entitlements but sure.

File: discord/enums.py [hunk 1]
```diff
@@ -1144,7 +1144,6 @@ class OnboardingPromptType(Enum):
     multiple_choice = 0
     dropdown        = 1
 
-
```
Comment (mccoderpy) [inline discord/enums.py:1147]: Why did you remove the blank line here? There should be 2 between top level classes/functions 👀

File: discord/enums.py [hunk 2]
```diff
@@ -1153,6 +1152,16 @@ class SKUType(Enum):
     subscription        = 5
     subscription_group  = 6
 
+class EntitlementType(Enum):
```
Comment (mccoderpy) [inline discord/enums.py:1158]: Please add this to the  `__all__` (public objects) tuple at the top of the file as well.

Comment (mccoderpy) [inline discord/enums.py:1158]: Also please document it at the bottom of [docs/api/enumes](https://github.com/mccoderpy/discord.py-message-components/blob/developer/docs/api/enums.rst?plain=1)

File: discord/monetization.py [hunk 3]
```diff
@@ -128,6 +140,12 @@ def __init__(self, data: m.EntitlementData, state: ConnectionState) -> None:
         self.deleted: bool = data['deleted']
         self.starts_at: datetime = utils.parse_time(data.get('starts_at'))
         self.ends_at: datetime = utils.parse_time(data.get('ends_at'))
+        self.consumed: bool = bool(data.get("consumed", None))
+        self._type = data["type"]
+
+    @property
+    def type(self) -> EntitlementType:
```
Comment (mccoderpy) [inline discord/monetization.py:146]:
```suggestion
    def type(self) -> EntitlementType:
        """:class:`discord.EntitlementType`: The type of the entitlement"""
```

File: discord/monetization.py [hunk 4]
```diff
@@ -105,6 +113,8 @@ class Entitlement:
     user_id: Optional[:class:`int`]
         :attr:`~discord.SKUFlags.app_user_subscription` only: The ID of the user that is granted access to the
         entitlement's sku
+    type: :class:`discord.EntitlementType`
```
Comment (mccoderpy) [inline discord/monetization.py]: Properties should only be documented in the property (getter) function itself, see suggestion bellow.

File: discord/monetization.py [hunk 5]
```diff
@@ -78,6 +78,14 @@ class SKU:
             This can be used to differentiate between a user and a server subscription.
     """
     __slots__ = ('id', 'type', 'application_id', 'name', 'slug', 'flags',)
+    
+    if TYPE_CHECKING:
```
Comment (mccoderpy) [inline discord/monetization.py]: This is duplicated as it's already type-hinted in the init

File: discord/monetization.py [hunk 6]
```diff
@@ -90,12 +98,12 @@ def __init__(self, data: m.SKU) -> None:
...
 class Entitlement:
     """
-    Entitlements represent a user's claim to a premium offering for your application.
+    Entitlements represent that a user or guild has access to a premium offering in your application.
 
     .. versionadded:: 2.0
 
     Attributes
-    ----------
+    =========
```
Comment (mccoderpy) [inline discord/monetization.py]:
```suggestion
    ---------
```

**Review by mccoderpy — CHANGES_REQUESTED** [review]
Please take a look at the comments and update the code accordingly

The rest looks fine so far, thank you for your contribution.

File: discord/enums.py (hunk 1 above)
Comment (The-LukeZ) [reply discord/enums.py:1147]: idk, I can readd it for sure

File: discord/enums.py (hunk 2 above)
Comment (The-LukeZ) [reply discord/enums.py:1158]: oh, yeah, forgot it

Comment (The-LukeZ) [reply discord/enums.py:1158]: ok

File: discord/enums.py (hunk 1 above)
Comment (mccoderpy) [reply discord/enums.py:1147]: Yea but we follow [PEP-8](https://peps.python.org/pep-0008/#blank-lines) when possible

File: discord/enums.py [hunk 7]
```diff
@@ -84,6 +84,7 @@
     'OnboardingMode',
     'OnboardingPromptType',
     'SKUType',
+    "EntitlementType",
```
Comment (mccoderpy) [inline discord/enums.py]:
```suggestion
    'EntitlementType',
```

File: discord/monetization.py [hunk 8]
```diff
@@ -90,12 +98,12 @@ def __init__(self, data: m.SKU) -> None:
...
 class Entitlement:
     """
-    Entitlements represent a user's claim to a premium offering for your application.
+    Entitlements represent that a user or guild has access to a premium offering in your application.
 
     .. versionadded:: 2.0
 
     Attributes
-    ----------
+    ---------
```
Comment (mccoderpy) [inline discord/monetization.py]:
```suggestion
    ----------
```

File: discord/monetization.py (hunk 4 above)
Comment (mccoderpy) [reply discord/monetization.py]: Please remove this as it is documented in the property

File: discord/monetization.py (hunk 5 above)
Comment (mccoderpy) [reply discord/monetization.py]: Please remove this :)

File: discord/monetization.py [hunk 9]
```diff
@@ -128,6 +140,12 @@ def __init__(self, data: m.EntitlementData, state: ConnectionState) -> None:
         self.deleted: bool = data['deleted']
         self.starts_at: datetime = utils.parse_time(data.get('starts_at'))
         self.ends_at: datetime = utils.parse_time(data.get('ends_at'))
+        self.consumed: bool = bool(data.get("consumed", None))
```
Comment (mccoderpy) [inline discord/monetization.py]:
```suggestion
        self.consumed: Optional[bool] = data.get('consumed')
```

**Review by mccoderpy — CHANGES_REQUESTED** [review]
Please remove the duplicated description/type-hints

File: discord/monetization.py [hunk 10]
```diff
@@ -128,6 +140,12 @@ def __init__(self, data: m.EntitlementData, state: ConnectionState) -> None:
         self.deleted: bool = data['deleted']
         self.starts_at: datetime = utils.parse_time(data.get('starts_at'))
         self.ends_at: datetime = utils.parse_time(data.get('ends_at'))
+        self.consumed: Optional[bool] = data.get("consumed")
```
Comment (mccoderpy) [inline discord/monetization.py]:
```suggestion
        self.consumed: Optional[bool] = data.get('consumed')
```

File: discord/monetization.py [hunk 11]
```diff
@@ -39,7 +39,7 @@
 from .types import monetization as m
 
 from . import utils
-from .enums import SKUType, try_enum
+from .enums import SKUType, EntitlementType, try_enum
 from .flags import SKUFlags
 
 if TYPE_CHECKING:
```
Comment (mccoderpy) [inline discord/monetization.py:45]:
```suggestion
if TYPE_CHECKING:
    from .user import User
    from .guild import Guild
```

File: discord/monetization.py [hunk 12]
```diff
@@ -128,6 +130,13 @@ def __init__(self, data: m.EntitlementData, state: ConnectionState) -> None:
...
+        self.consumed: Optional[bool] = data.get('consumed')
+        self._type: int = data["type"]
+
+    @property
+    def type(self) -> EntitlementType:
+        """:class:`discord.EntitlementType`: The type of the entitlement"""
+        return try_enum(EntitlementType, self._type)
 
     @property
     def target(self):
```
Comment (mccoderpy) [inline discord/monetization.py]:
```suggestion
    def target(self) -> Union[User, Guild]:
```

File: discord/monetization.py [hunk 13]
```diff
@@ -128,6 +132,13 @@ def __init__(self, data: m.EntitlementData, state: ConnectionState) -> None:
         self.deleted: bool = data['deleted']
         self.starts_at: datetime = utils.parse_time(data.get('starts_at'))
         self.ends_at: datetime = utils.parse_time(data.get('ends_at'))
+        self.consumed: Optional[bool] = data.get('consumed')
+        self._type: int = data["type"]
+
+    @property
+    def type(self) -> EntitlementType:
+        """:class:`discord.EntitlementType`: The type of the entitlement"""
```
Comment (mccoderpy) [inline discord/monetization.py]:
```suggestion
        """:class:`~discord.EntitlementType`: The type of the entitlement"""
```

File: discord/monetization.py [hunk 14]
```diff
@@ -114,9 +116,11 @@ class Entitlement:
         Start date at which the entitlement is valid. ``None`` for test entitlements.
     ends_at: Optional[:class:`datetime.datetime`]
         Time at which the entitlement is no longer valid. ``None`` for test entitlements.
-
+    consumed: Optional[:class:`bool`]
+            Indicates whether the entitlement has been consumed for consumable items.
+            Can be ``None`` in certain cases.
```
Comment (mccoderpy) [inline discord/monetization.py]:
```suggestion

```
That it is `None` for non-consumables is already clear by the first sentence.

**Review by mccoderpy — APPROVED** [review]

**Pull request merged**